FurkannOzbek/application-hub
# [FEATURE] Implement Notes Management System for Applications

# [FEATURE] Implement Notes Management System for Applications 

---

## Description

This PR implements a complete notes management system for applications, allowing users to create, edit, delete, and view notes associated with specific applications. The implementation includes both frontend and backend changes.

**Problem Solved:**
1. Previously, there was no way to manage notes for applications

**Impact of Change:** 
1. Users can now maintain detailed notes for each application
2. Improved user experience with rich text editing capabilities


---

## Changes Made

### Frontend Changes
- Notes Component (`frontend/app/applications/[id]/tabs/Notes.js`)
   - Implemented full CRUD operations for notes
   - Added rich text editor integration
   - Implemented loading states and error handling
   - Added delete confirmation modal
   - Improved UI with proper button states and layouts

- Loader Component (`frontend/components/ui/Loader.js`)
   - Added customizable height prop for flexible loading states

- API Utils (`frontend/utils/api.js`)
   - Added new API endpoints for notes:
      - `getNoteByApplicationId`
      - `createOrUpdateNote`
      - `deleteNote`

#### Backend Changes
- Note Controller (`api/src/controllers/noteController.js`)
   - Modified note retrieval logic to return empty note structure when no note exists


---

## Screenshots 

<div>
    <a href="https://www.loom.com/share/5f5ef134991249d1a6d739d047521019">
      <p>ApplicationHub App - 27 November 2024 - Watch Video</p>
    </a>
    <a href="https://www.loom.com/share/5f5ef134991249d1a6d739d047521019">
      <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/5f5ef134991249d1a6d739d047521019-f2db253f49b44731-full-play.gif">
    </a>
  </div>


https://github.com/user-attachments/assets/2cdebcff-5574-48d1-9ded-6dbd47377626


---

## Testing Steps

1. Navigate to an application's notes tab
2. Test note creation:
 - Click "Add" button
 - Enter note content
 - Save the note

2. Test note editing:
 - Click "Edit" button
 - Modify content
 - Save changes

3. Test note deletion:
 - Click "Delete" button
 - Confirm deletion in modal

5. Verify loading states and error handling

---

## Issues Addressed

Closes [Enhance UI for Notes + use components  + add functionality](https://trello.com/c/ct8g5NwZ)
